derive/src/traits.rs

```diff
@@ -981,14 +1005,32 @@ fn generate_checked_bit_pattern_enum_with_fields(
   }
 }
 
-/// Check that a struct has no padding by asserting that the size of the struct
-/// is equal to the sum of the size of it's fields
-fn generate_assert_no_padding(input: &DeriveInput) -> Result<TokenStream> {
+/// Check that a struct or enum has no padding by asserting that the size of
+/// the type is equal to the sum of the size of it's fields and enum
+/// discriminant
+fn generate_assert_no_padding(
+  input: &DeriveInput, enum_variant: Option<&Variant>,
+) -> Result<TokenStream> {
   let struct_type = &input.ident;
-  let enum_variant = None; // `no padding` check is not supported for `enum`s yet.
   let fields = get_fields(input, enum_variant)?;
 
-  let mut field_types = get_field_types(&fields);
+  // If the type is an enum, determine the type of its discriminant.
+  let enum_discriminant = if matches!(input.data, Data::Enum(_)) {
+    let repr = get_repr(&input.attrs)?;
+    let integer = match repr.repr {
+      Repr::C => quote!(::core::ffi::c_int),
+      Repr::CWithDiscriminant(integer) => quote!(#integer),
+      _ => unreachable!(),
```

Comment:
```suggestion
      _ => bail!("Deriving NoUninit is not currently supported for repr(Int) enums with fields")
```

`#[repr(Int)]` enums also have a ([different](https://doc.rust-lang.org/reference/type-layout.html#primitive-representation-of-enums-with-fields)) defined layout to `#[repr(C)]`/`#[repr(C, Int)]`, so ideally those would also be allowed, but if you don't want to add that in this PR, then this should probably be a `bail!` with an error message instead of `unreachable!`.

E.g. this `unreachable!()` currently causes a "proc-macro derive panicked" for a `#[repr(u8)]` enum.

```rs
use bytemuck::NoUninit;

#[derive(Debug, Clone, Copy, NoUninit)]
#[repr(u8)]
enum Foo {
    A(u8),
}
```

```rs
error: proc-macro derive panicked
 --> src/lib.rs:3:30
  |
3 | #[derive(Debug, Clone, Copy, NoUninit)]
  |                              ^^^^^^^^
  |
  = help: message: internal error: entered unreachable code
```

Reply:
Good catch, thanks! AFAICT the layout is the same if there is no padding, so adding support for this was easy. I added a test for `#[repr(Int)]`.